File: tests/sdk/nodejs/identity.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { doJSON, hasOwn, projectID, requireReady, requireToken, uniqueID } from "./helpers.mjs";

test("ready and get project smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON("GET", `${ep}/v3/projects/${project}`, token);
  assert.equal(status, 200, `get project status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.equal(parsed.projectId, project, body);
});

test("get organization smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;

  const { status, body } = await doJSON(
    "GET",
    `${ep}/v3/organizations/noctaxris-gcp-org`,
    token,
  );
  assert.equal(status, 200, `get organization status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.equal(parsed.name, "organizations/noctaxris-gcp-org", body);
});

test("list folders smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;

  const parent = encodeURIComponent("organizations/noctaxris-gcp-org");
  const { status, body } = await doJSON("GET", `${ep}/v3/folders?parent=${parent}`, token);
  assert.equal(status, 200, `list folders status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "folders"), body);
});

test("iam generateAccessToken smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();
  let accountId = uniqueID("sdksa");
  if (accountId.length > 30) {
    accountId = accountId.slice(0, 30).replace(/-+$/, "");
  }
  const email = `${accountId}@${project}.iam.gserviceaccount.com`;
  const saPath = `${ep}/v1/projects/${project}/serviceAccounts/${email}`;

  const created = await doJSON("POST", `${ep}/v1/projects/${project}/serviceAccounts`, token, {
    accountId,
    serviceAccount: { displayName: "sdk smoke" },
  });
  assert.equal(created.status, 200, `create service account status=${created.status} body=${created.body}`);
  t.after(async () => {
    try {
      await doJSON("DELETE", saPath, token);
    } catch {
      /* best-effort cleanup */
    }
  });

  const { status, body } = await doJSON("POST", `${saPath}:generateAccessToken`, token, {
    scope: ["https://www.googleapis.com/auth/cloud-platform"],
    lifetime: "3600s",
  });
  assert.equal(status, 200, `generateAccessToken status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(parsed.accessToken, body);
});

test("sts token exchange smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();
  let poolId = uniqueID("sdk-sts-pool");
  if (poolId.length > 32) {
    poolId = poolId.slice(0, 32).replace(/-+$/, "");
  }
  const providerId = "oidc";
  const poolBase = `${ep}/v1/projects/${project}/locations/global/workloadIdentityPools/${poolId}`;
  const providerName = `projects/${project}/locations/global/workloadIdentityPools/${poolId}/providers/${providerId}`;

  const pool = await doJSON(
    "POST",
    `${ep}/v1/projects/${project}/locations/global/workloadIdentityPools?workloadIdentityPoolId=${encodeURIComponent(poolId)}`,
    token,
    { displayName: "sdk sts pool" },
  );
  assert.equal(pool.status, 200, `create WIF pool status=${pool.status} body=${pool.body}`);
  t.after(async () => {
    try {
      await doJSON("DELETE", poolBase, token);
    } catch {
      /* best-effort cleanup */
    }
  });

  const provider = await doJSON(
    "POST",
    `${poolBase}/providers?workloadIdentityPoolProviderId=${encodeURIComponent(providerId)}`,
    token,
    { displayName: "sdk oidc", oidc: { issuerUri: "https://example.com" } },
  );
  assert.equal(provider.status, 200, `create WIF provider status=${provider.status} body=${provider.body}`);
  t.after(async () => {
    try {
      await doJSON("DELETE", `${poolBase}/providers/${providerId}`, token);
    } catch {
      /* best-effort cleanup */
    }
  });

  const form = new URLSearchParams({
    grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
    audience: `//iam.googleapis.com/${providerName}`,
    subject_token: "sdk-sts-sub",
    subject_token_type: "urn:ietf:params:oauth:token-type:jwt",
  });
  const res = await fetch(`${ep}/v1/token`, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: form.toString(),
    signal: AbortSignal.timeout(5000),
  });
  const body = await res.text();
  assert.equal(res.status, 200, `STS /v1/token status=${res.status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(parsed.access_token, body);
  assert.equal(parsed.token_type, "Bearer", body);
});

test("list service usage services smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON("GET", `${ep}/v1/projects/${project}/services`, token);
  assert.equal(status, 200, `list Service Usage services status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "services"), body);
});

File: tests/sdk/go/identity_test.go
package sdk_test

import (
	"encoding/json"
	"net/http"
	"net/url"
	"strings"
	"testing"
)

func TestReadyAndGetProjectSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	status, body := doJSON(t, http.MethodGet, ep+"/v3/projects/"+project, token, nil)
	if status != http.StatusOK {
		t.Fatalf("get project status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode project: %v body=%s", err, body)
	}
	if got, _ := parsed["projectId"].(string); got != project {
		t.Fatalf("projectId=%v want %s body=%s", parsed["projectId"], project, body)
	}
}

func TestGetOrganizationSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)

	status, body := doJSON(t, http.MethodGet, ep+"/v3/organizations/noctaxris-gcp-org", token, nil)
	if status != http.StatusOK {
		t.Fatalf("get organization status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode organization: %v body=%s", err, body)
	}
	if got, _ := parsed["name"].(string); got != "organizations/noctaxris-gcp-org" {
		t.Fatalf("name=%v want organizations/noctaxris-gcp-org body=%s", parsed["name"], body)
	}
}

func TestListFoldersSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)

	q := url.Values{"parent": {"organizations/noctaxris-gcp-org"}}
	status, body := doJSON(t, http.MethodGet, ep+"/v3/folders?"+q.Encode(), token, nil)
	if status != http.StatusOK {
		t.Fatalf("list folders status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode folders: %v body=%s", err, body)
	}
	if _, ok := parsed["folders"]; !ok {
		t.Fatalf("missing folders field body=%s", body)
	}
}

func TestIAMGenerateAccessTokenSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()
	accountID := uniqueID("sdksa")
	if len(accountID) > 30 {
		accountID = accountID[:30]
		accountID = strings.TrimRight(accountID, "-")
	}
	email := accountID + "@" + project + ".iam.gserviceaccount.com"
	saPath := ep + "/v1/projects/" + project + "/serviceAccounts/" + email

	status, body := doJSON(t, http.MethodPost, ep+"/v1/projects/"+project+"/serviceAccounts", token, map[string]any{
		"accountId": accountID,
		"serviceAccount": map[string]any{
			"displayName": "sdk smoke",
		},
	})
	if status != http.StatusOK {
		t.Fatalf("create service account status=%d body=%s", status, body)
	}
	t.Cleanup(func() {
		_, _, _ = doJSONErr(http.MethodDelete, saPath, token, nil)
	})

	status, body = doJSON(t, http.MethodPost, saPath+":generateAccessToken", token, map[string]any{
		"scope":    []string{"https://www.googleapis.com/auth/cloud-platform"},
		"lifetime": "3600s",
	})
	if status != http.StatusOK {
		t.Fatalf("generateAccessToken status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode generateAccessToken: %v body=%s", err, body)
	}
	if got, _ := parsed["accessToken"].(string); got == "" {
		t.Fatalf("missing accessToken body=%s", body)
	}
}

func TestSTSTokenExchangeSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()
	poolID := uniqueID("sdk-sts-pool")
	if len(poolID) > 32 {
		poolID = poolID[:32]
		poolID = strings.TrimRight(poolID, "-")
	}
	providerID := "oidc"
	poolBase := ep + "/v1/projects/" + project + "/locations/global/workloadIdentityPools/" + poolID
	providerName := "projects/" + project + "/locations/global/workloadIdentityPools/" + poolID + "/providers/" + providerID

	status, body := doJSON(t, http.MethodPost, ep+"/v1/projects/"+project+"/locations/global/workloadIdentityPools?workloadIdentityPoolId="+url.QueryEscape(poolID), token, map[string]any{
		"displayName": "sdk sts pool",
	})
	if status != http.StatusOK {
		t.Fatalf("create WIF pool status=%d body=%s", status, body)
	}
	t.Cleanup(func() {
		_, _, _ = doJSONErr(http.MethodDelete, poolBase, token, nil)
	})

	status, body = doJSON(t, http.MethodPost, poolBase+"/providers?workloadIdentityPoolProviderId="+url.QueryEscape(providerID), token, map[string]any{
		"displayName": "sdk oidc",
		"oidc":        map[string]any{"issuerUri": "https://example.com"},
	})
	if status != http.StatusOK {
		t.Fatalf("create WIF provider status=%d body=%s", status, body)
	}
	t.Cleanup(func() {
		_, _, _ = doJSONErr(http.MethodDelete, poolBase+"/providers/"+providerID, token, nil)
	})

	form := url.Values{
		"grant_type":         {"urn:ietf:params:oauth:grant-type:token-exchange"},
		"audience":           {"//iam.googleapis.com/" + providerName},
		"subject_token":      {"sdk-sts-sub"},
		"subject_token_type": {"urn:ietf:params:oauth:token-type:jwt"},
	}
	status, raw, err := doForm(http.MethodPost, ep+"/v1/token", "application/x-www-form-urlencoded", form.Encode())
	if err != nil {
		t.Fatalf("STS /v1/token: %v", err)
	}
	if status != http.StatusOK {
		t.Fatalf("STS /v1/token status=%d body=%s", status, raw)
	}
	var parsed map[string]any
	if err := json.Unmarshal(raw, &parsed); err != nil {
		t.Fatalf("decode STS: %v body=%s", err, raw)
	}
	if got, _ := parsed["access_token"].(string); got == "" {
		t.Fatalf("missing access_token body=%s", raw)
	}
	if got, _ := parsed["token_type"].(string); got != "Bearer" {
		t.Fatalf("token_type=%v want Bearer body=%s", parsed["token_type"], raw)
	}
}

func TestListServiceUsageServicesSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	path := ep + "/v1/projects/" + project + "/services"
	status, body := doJSON(t, http.MethodGet, path, token, nil)
	if status != http.StatusOK {
		t.Fatalf("list Service Usage services status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode Service Usage services: %v body=%s", err, body)
	}
	if _, ok := parsed["services"]; !ok {
		t.Fatalf("missing services field body=%s", body)
	}
}
